### PR TITLE
Add validation tests for trace XSD

### PR DESCRIPTION
**Pull request by hannes-steffenhagen-diffblue**
We have a XSD for cbmc --trace --xml-ui in our documentation, this adds
a new test run which essentially validates the XML output of all tests
in regression/cbmc against that XSD. Not all tests in there actually
produce traces, but most do so we just exclude the few that don't.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by NlightNFotis — COMMENTED** [review]
LGTM, but cannot comment on the gradle build files and the .bat files.

File: regression/validate-trace-xml-schema/validate-xsd/gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
```
Comment (hannes-steffenhagen-diffblue) [inline regression/validate-trace-xml-schema/validate-xsd/gradle/wrapper/gradle-wrapper.properties]: Everything in the `gradle` directory as well as the `gradlew` and `gradlew.bat` scripts are autogenerated by gradle so the build can work without a user having to have a compatible gradle version locally. So they're meant to be here, but shouldn't be reviewed.

File: regression/Makefile [hunk 2]
```diff
@@ -2,35 +2,36 @@
 # running tests in parallel, it is important that these directories are
 # listed with decreasing runtimes (i.e. longest running at the top)
 DIRS = cbmc \
-       cbmc-library \
```
Comment (thk123) [inline regression/Makefile]: :question: I _assume_ this is just a reordering of the Makefile? If it is, it'd be best pulled out into its own commit so can be clearly stated as such

File: regression/validate-trace-xml-schema/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+if(${PYTHONINTERP_FOUND}
+  AND ${Java_FOUND} AND ${Java_VERSION_MAJOR} VERSION_GREATER 8)
+    add_test(NAME validate-trace-xml-schema
+            COMMAND ${PYTHON_EXECUTABLE} check.py $<TARGET_FILE:cbmc>
+            WORKING_DIRECTORY ${CMAKE_CURRENT_LIST_DIR})
+    set_tests_properties(validate-trace-xml-schema
+            PROPERTIES
+            LABELS "CBMC;CORE")
+else()
+    message(WARNING "JDK not found, skipping trace schema tests")
```
Comment (thk123) [inline regression/validate-trace-xml-schema/CMakeLists.txt]: :question: This stops the tests being configured? I wonder if we should configure a test that always fails? That way, if someone configures it locally, they don't need JDK, but if CI fails on this step, it doesn't silently pass.

File: regression/validate-trace-xml-schema/check.sh [hunk 4]
```diff
@@ -0,0 +1,29 @@
+if ! command -v python3 > /dev/null; then
+    echo "python3 not found, skipping XSD tests"
+    exit
+fi
+
+python_minor_version=$(python3 --version | cut -d ' ' -f 2 | cut -d '.' -f 2)
+
+if [ $python_minor_version -lt 5 ]; then
+    echo "python version less than 3.5, skipping XSD tests"
```
Comment (thk123) [inline regression/validate-trace-xml-schema/check.sh:9]: :bulb: I think I would exit with a non-zero exit code here, to prevent silent failures on CI :ghost:

File: regression/validate-trace-xml-schema/check.sh [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    exit
+fi
+
+if ! command -v javac > /dev/null; then
+    echo "javac not found, skipping XSD tests"
+    exit
```
Comment (thk123) [inline regression/validate-trace-xml-schema/check.sh]: :ghost:

File: regression/validate-trace-xml-schema/check.sh [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+# Java versions are complicated
+
+java_version=$(javac -version | cut -d ' ' -f 2)
+javac_major_version=$(echo $java_version | cut -d '.' -f 1)
+javac_minor_version=$(echo $java_version | cut -d '.' -f 2)
+
+if [ $javac_major_version -eq 1 ] || [ $javac_minor_version -ge 8 ]; then
+    echo "JDK version less than 8, skipping XSD tests"
+    exit
```
Comment (thk123) [inline regression/validate-trace-xml-schema/check.sh]: :ghost:

File: regression/validate-trace-xml-schema/validate-xsd/src/main/java/com/diffblue/validatexsd/Main.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import javax.xml.validation.Schema;
+import javax.xml.validation.SchemaFactory;
+import javax.xml.validation.Validator;
+import java.io.FileInputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.concurrent.Callable;
+
+@Command(name = "validate-xsd", mixinStandardHelpOptions = true, version = "0.1",
```
Comment (thk123) [inline regression/validate-trace-xml-schema/validate-xsd/src/main/java/com/diffblue/validatexsd/Main.java]: It makes me a little sad we weren't able to find an xsd validator that exists on apt, but there is not much code here anyway so I guess it's OK

File: regression/validate-trace-xml-schema/validate-xsd/src/main/java/com/diffblue/validatexsd/Main.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  description = "tells you if given XML (read from stdin) matches a given XSD spec")
```
Comment (thk123) [inline regression/validate-trace-xml-schema/validate-xsd/src/main/java/com/diffblue/validatexsd/Main.java]: :bulb: An example usage would be great:

```
Usage:

$ cbmc test.c --trace | java -jar validator.jar 
```

File: regression/validate-trace-xml-schema/validate-xsd/src/main/java/com/diffblue/validatexsd/Main.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+public class Main implements Callable<Integer> {
+  @Parameters(index = "0")
+  private Path xsdSchemaFilePath;
+
+  @Override
+  public Integer call() throws Exception {
```
Comment (thk123) [inline regression/validate-trace-xml-schema/validate-xsd/src/main/java/com/diffblue/validatexsd/Main.java]: :no_entry_sign: Sorry to be a pain, but can we have 4 simple tests just to check the errors get thrown correctly:

 - no schema
 - invalid schema
 - invalid xml
 - valid xml

Hopefully these don't take long to write.

File: regression/validate-trace-xml-schema/validate-xsd/src/main/java/com/diffblue/validatexsd/LoggingErrorHandler.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+import org.xml.sax.ErrorHandler;
+import org.xml.sax.SAXException;
+import org.xml.sax.SAXParseException;
+
+public class LoggingErrorHandler implements ErrorHandler {
+  private boolean errorOccurred = false;
+
+  @Override
+  public void warning(SAXParseException e) throws SAXException {
```
Comment (thk123) [inline regression/validate-trace-xml-schema/validate-xsd/src/main/java/com/diffblue/validatexsd/LoggingErrorHandler.java]: Should you be re-throwing these execptions? Seems weird the signature would throw otherwise?

File: regression/validate-trace-xml-schema/check.py [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+this_script_dir = os.path.abspath(os.path.dirname(__file__))
+
+TraceXsdSpec = os.path.abspath(os.path.join(this_script_dir, '..', '..', 'doc/assets/xml_spec.xsd'))
+
+test_base_dir = os.path.abspath(os.path.join(this_script_dir, '..', 'cbmc'))
+
+# some tests in the cbmc suite don't work for the trace checks for one reason or another
+ExcludedTests = list(map(lambda s: os.path.join(test_base_dir, s), [
+    # these tests except input from stdin
```
Comment (thk123) [inline regression/validate-trace-xml-schema/check.py:15]: expect? accept? I don't think you mean except here though

File: regression/validate-trace-xml-schema/check.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    # these want --show-goto-functions instead of producing a trace
+    'destructors/compound_literal.desc',
+    'destructors/enter_lexical_block.desc',
+    'reachability-slice-interproc2/test.desc',
+    # this one wants show-properties instead producing a trace
+    'show_properties1/test.desc',
+    # program-only instead of trace
+    'vla1/program-only.desc',
+    # this one doesn't work for me locally at all
+    'integer-assignments1/test.desc',
```
Comment (thk123) [inline regression/validate-trace-xml-schema/check.py]: :question: Huh? You mean the test fails locally? Perhaps we could raise an issue to look into that?

File: regression/validate-trace-xml-schema/check.py [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+    # these test for invalid command line handling
+    'bad_option/test_multiple.desc',
+    'bad_option/test.desc',
+    # this one produces XML intermingled with main XML output when used with --xml-ui
+    'graphml_witness2/test.desc',
+    # this one uses program-only again
+    'Quantifiers-simplify/simplify_not_forall.desc',
```
Comment (thk123) [inline regression/validate-trace-xml-schema/check.py]: :bulb: I guess these are in alphabetical order, but might be better to group by reason?

File: regression/validate-trace-xml-schema/check.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+    'reachability-slice/test.desc',
+    'reachability-slice/test2.desc',
+    'reachability-slice/test3.desc',
+    'reachability-slice-interproc/test.desc',
+    # again, intermingles several XML documents
+    'graphml_witness1/test.desc',
+    'switch8/program-only.desc',
+    # show-goto-functions
+    'unsigned1/test.desc',
+    # show-goto-functions
```
Comment (thk123) [inline regression/validate-trace-xml-schema/check.py]: :pick: Combine these groups

File: regression/validate-trace-xml-schema/check.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+    def __exit__(self, _type, _value, _tb):
+        os.chdir(self.current_wd)
+
+
+class TestSpec:
+    def __init__(self, args, is_knownbug, is_future, is_thorough):
+        self.args = args
+        self.is_knownbug = is_knownbug
+        self.is_future = is_future
+        self.is_thorough = is_thorough
```
Comment (thk123) [inline regression/validate-trace-xml-schema/check.py:83]: I think we've written a `.desc` python parser before (possibly for test-gen) :bulb: Pull out that into a separate python module so that it may be reused

File: regression/validate-trace-xml-schema/check.py [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+            print('[SUCCESS]')
+        else:
+            print('[FAILURE]')
+            self.failed_tests.append(spec_path)
+        sys.stdout.buffer.write(validate_result.stdout)
+        sys.stdout.buffer.write(validate_result.stderr)
+
+    def read_spec(self, spec_path):
+        with open(spec_path) as spec_file:
+            spec = spec_file.readline().split()
```
Comment (thk123) [inline regression/validate-trace-xml-schema/check.py]: :bulb: I guess I would use the word `test_kind` rather than spec, which has 3 means in this file:
1. the XSD specificaiton
2. the entire .desc file
3. The first line of the desc file

File: regression/validate-trace-xml-schema/check.py [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+                                         stderr=subprocess.PIPE)
+        if validate_result.returncode == 0:
+            print('[SUCCESS]')
+        else:
+            print('[FAILURE]')
+            self.failed_tests.append(spec_path)
+        sys.stdout.buffer.write(validate_result.stdout)
+        sys.stdout.buffer.write(validate_result.stderr)
+
+    def read_spec(self, spec_path):
```
Comment (thk123) [inline regression/validate-trace-xml-schema/check.py]: :bulb: Here your spec is the `.desc` file right? I would call this `read_test_desc`

**Review by thk123 — CHANGES_REQUESTED** [review]
This seems good - some naming stuff in the python file I'd really like to see tidied up. Also, I sadly must insist on some tests for the program than runs the validation.

File: regression/validate-trace-xml-schema/CMakeLists.txt (hunk 3 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/validate-trace-xml-schema/CMakeLists.txt]: Not all our CI passes have both Java and Python 3 available.

File: regression/validate-trace-xml-schema/check.sh (hunk 4 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/validate-trace-xml-schema/check.sh:9]: This is not supposed to fail CI

File: regression/validate-trace-xml-schema/validate-xsd/src/main/java/com/diffblue/validatexsd/Main.java (hunk 7 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/validate-trace-xml-schema/validate-xsd/src/main/java/com/diffblue/validatexsd/Main.java]: So after doing some more digging apparently there's xmllint? I think I can remove the java bits and just use that.

File: regression/validate-trace-xml-schema/validate-xsd/src/main/java/com/diffblue/validatexsd/LoggingErrorHandler.java (hunk 10 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/validate-trace-xml-schema/validate-xsd/src/main/java/com/diffblue/validatexsd/LoggingErrorHandler.java]: You can rethrow, but you don't have to.

File: regression/validate-trace-xml-schema/validate-xsd/src/main/java/com/diffblue/validatexsd/Main.java (hunk 7 above)
Comment (thk123) [reply regression/validate-trace-xml-schema/validate-xsd/src/main/java/com/diffblue/validatexsd/Main.java]: Sounds good -> saves writing the tests! :tada:

File: regression/validate-trace-xml-schema/CMakeLists.txt (hunk 3 above)
Comment (thk123) [reply regression/validate-trace-xml-schema/CMakeLists.txt]: I guess I'd rather we explicitly _skip_ this test on the jobs that don't have the requirements.

**Comment by thk123** [comment]
I've checked the test appears in the output for one job, but just to be safe, could you break the output trace to ensure the CI fails (e.g. just change one of the fields or something so is nonsense, ideally one that doesn't cause any other tests to fail) (maybe add a field?)

File: regression/validate-trace-xml-schema/check.py (hunk 11 above)
Comment (thk123) [reply regression/validate-trace-xml-schema/check.py:15]: > expect? accept? I don't think you mean except here though

:pick: This seems outstanding

File: regression/validate-trace-xml-schema/check.py [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+    'destructors/compound_literal.desc',
+    'destructors/enter_lexical_block.desc',
+    'reachability-slice-interproc2/test.desc',
+    # this one wants show-properties instead producing a trace
+    'show_properties1/test.desc',
+    # program-only instead of trace
+    'vla1/program-only.desc',
+    'Quantifiers-simplify/simplify_not_forall.desc',
+    # this one doesn't work for me locally at all
+    # 'integer-assignments1/test.desc',
```
Comment (thk123) [inline regression/validate-trace-xml-schema/check.py:32]: >  Huh? You mean the test fails locally? Perhaps we could raise an issue to look into that?

:question: outstanding, this needs to be addressed IMO

File: regression/validate-trace-xml-schema/check.sh (hunk 4 above)
Comment (thk123) [reply regression/validate-trace-xml-schema/check.sh:9]: What I'm suggesting is it should fail CI (and the CIs where this doesn't work, we disable it in the configuration file). Currently if a config breaks in a way that means we have the wrong version of python, this test will just be disabled and we'll never know.

File: regression/validate-trace-xml-schema/check.py (hunk 16 above)
Comment (thk123) [reply regression/validate-trace-xml-schema/check.py]: Would like to see this addressed, as I keep thinking `is_knownbug = 'KNOWNBUG' in spec` is a bug, since KNOWNBUG might appear elsewhere in the spec, except here `spec` is just the first line.

Where "this" is:

Rename `spec` to `spec_kind` since `spec` makes it seem like it is the whole contents of the `desc` file.

**Review by thk123 — COMMENTED** [review]
One remaining blocking comment (answering my question about the test you've labelled as failing locally).

I also strongly believe scripts that we run in CI should silently fail (e.g. when python not installed)

File: regression/validate-trace-xml-schema/check.py (hunk 18 above)
Comment (thk123) [reply regression/validate-trace-xml-schema/check.py:32]: To be explicit about what needs addressing:
 - shouldn't check in "me" since no way of knowing who "me" is as soon as this line gets moved. 

Further, I'd like to understand what you mean "doesn't work locally at all" means!

File: regression/validate-trace-xml-schema/check.py (hunk 15 above)
Comment (thk123) [reply regression/validate-trace-xml-schema/check.py:83]: For reference, we [did](https://github.com/diffblue/cover/blob/develop/test-gen/regression/common.py#L50 :lock:),  but not trivial to unify these two at the moment.

**Review by thk123 — APPROVED** [review]
I give up on trying to get the CI to fail without dependencies:
 - for make builds on windows, can't get all the dependencies, and `make test` doesn't offer the opportunity to filter _out_ tests AFAICT
 - for cmake builds, though you can filter out the tests, the info about what dependency gets lost which makes it a pain to debug, so I'll leave it

I'll raise a separate PR addressing my other comments, since I've fixed them, but let's get this in first (needs a rebase which is probably easier for you to do?

**Review by peterschrammel — APPROVED** [review]
